plugins/check-path/core-rs/src/paths.rs

```diff
@@ -0,0 +1,84 @@
+use js_sys::Reflect;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+use wasm_bindgen::prelude::*;
+
+enum NodeValue {
+    StringType(String),
+    ObjectType(JsValue),
+    None,
+}
+
+pub struct Paths {
+    store: Rc<RefCell<HashMap<String, Vec<String>>>>,
+}
+
+impl Paths {
+    pub fn new() -> Self {
+        Self {
+            store: Rc::new(RefCell::new(HashMap::new())),
+        }
+    }
+
+    pub fn parse(&self, data: JsValue) {
+        self.traverse(data);
+    }
+
+    pub fn get(&self, key: &str) -> Vec<String> {
+        self.store
+            .borrow()
+            .get(key)
+            .or(Some(&vec!["".to_string()]))
+            .unwrap()
+            .clone()
+    }
+
+    fn traverse(&self, root: JsValue) {
+        let path: Vec<String> = vec![];
+        let mut stack =
+            Paths::get_key_values(&root, path.clone()).expect("Couldn't read root node.");
+        while let Some((key, value, path)) = stack.pop() {
+            if let NodeValue::StringType(value) = value {
+                if key == "id" || key == "type" {
+                    self.store.borrow_mut().insert(value, path);
+                }
+            } else if let NodeValue::ObjectType(value) = value {
+                let mut object_path = path.clone();
+                object_path.push(key);
+                stack.append(Paths::get_key_values(&value, object_path).unwrap().as_mut());
+            }
+        }
+    }
+
+    fn get_key_values(
+        value: &JsValue,
+        path: Vec<String>,
+    ) -> Option<Vec<(String, NodeValue, Vec<String>)>> {
```

Comment:
```suggestion
    ) -> impl Iterator<Item=(String, NodeValue, Vec<String>)> {
```
Or something like this.

Then return the iterator instead of collecting it into a vec.

Reply:
just got rid of this function `get_key_values`, iterating through all keys here just to add to stack to iterate again was not the best approach.